### PR TITLE
Test framework fixes and retry count mechanism

### PR DESCRIPTION
**Pull request by soninaren**

### CODE REVIEW

File: e2etest/src/main/java/com/microsoft/windowsazure/mobileservices/zumoe2etestapp/framework/TestCase.java [hunk 1]
```diff
@@ -101,6 +102,18 @@ public void setEnabled(boolean enabled) {
...
+    public void setRetryCount(int retryCount) {
+        if (!isRetryCountSet) {
+            mtryCount  = mtryCount + retryCount;
+            isRetryCountSet = true;
+        }
+    }
+
+    public int retryCount() {
+        return mtryCount - 1;
+    }
```
Comment (mamaso) [inline e2etest/src/main/java/com/microsoft/windowsazure/mobileservices/zumoe2etestapp/framework/TestCase.java]: nit: retryCount getter seems unnecessary

File: e2etest/src/main/java/com/microsoft/windowsazure/mobileservices/zumoe2etestapp/framework/TestCase.java [hunk 2]
```diff
@@ -38,6 +38,7 @@
 
     private boolean mCanRunUnattended;
 
+    private int mtryCount = 1;
```
Comment (mamaso) [inline e2etest/src/main/java/com/microsoft/windowsazure/mobileservices/zumoe2etestapp/framework/TestCase.java]: nit: mTryCount

File: e2etest/src/main/java/com/microsoft/windowsazure/mobileservices/zumoe2etestapp/framework/TestCase.java [hunk 3]
```diff
@@ -117,28 +130,32 @@ public void run(MobileServiceClient client, final TestExecutionCallback callback
...
 
-            });
+                    @Override
+                    public void onTestGroupComplete(TestGroup group, List<TestResult> results) {
+                        // This will never be called
+                    }
+
+                });
+                mtryCount--;
+            }
```
Comment (mamaso) [inline e2etest/src/main/java/com/microsoft/windowsazure/mobileservices/zumoe2etestapp/framework/TestCase.java]: Perhaps add logging for test retries?

File: e2etest/src/main/java/com/microsoft/windowsazure/mobileservices/zumoe2etestapp/framework/TestCase.java [hunk 4]
```diff
@@ -117,28 +130,32 @@ public void run(MobileServiceClient client, final TestExecutionCallback callback
...
-                public void onTestGroupComplete(TestGroup group, List<TestResult> results) {
-                    // This will never be called
-                }
+                    @Override
+                    public void onTestComplete(TestCase test, TestResult result) {
+                        thisTest.mEndTime = new Date();
+                        if (result != null && result.getTestCase() == null) {
+                            result.setTestCase(thisTest);
+                        }
+                        callback.onTestComplete(test, result);
```
Comment (mamaso) [inline e2etest/src/main/java/com/microsoft/windowsazure/mobileservices/zumoe2etestapp/framework/TestCase.java]: Do we want to execute the callback.onTestComplete callback multiple times when retrying?

Comment (soninaren) [inline e2etest/src/main/java/com/microsoft/windowsazure/mobileservices/zumoe2etestapp/framework/TestCase.java]: yes :)

File: e2etest/src/main/java/com/microsoft/windowsazure/mobileservices/zumoe2etestapp/framework/TestCase.java (hunk 1 above)
Comment (soninaren) [inline e2etest/src/main/java/com/microsoft/windowsazure/mobileservices/zumoe2etestapp/framework/TestCase.java]: fixed

File: e2etest/src/main/java/com/microsoft/windowsazure/mobileservices/zumoe2etestapp/framework/TestCase.java (hunk 2 above)
Comment (soninaren) [inline e2etest/src/main/java/com/microsoft/windowsazure/mobileservices/zumoe2etestapp/framework/TestCase.java]: fixed

**Comment by mamaso** [comment]
Looks good, merge away

**Pull request merged**